File: web-service/mongo_models/DependencyGraph.js
const mongoose = require('mongoose');


const UtilizedEntitySchema = new mongoose.Schema({
    name: { type: String, required: true },
    type: { type: String, required: true },
    alias: { type: String },
    source: { type: String, enum: ['Exporting', 'Importing'], required: true }
});

const DependentNodeSchema = new mongoose.Schema({
    name: { type: String, required: true },
    alias: { type: String },
    valid: { type: Boolean, required: true },
    weight: { type: [UtilizedEntitySchema], default: [] }
});

// const FileNodeSchema = new mongoose.Schema({
//     name: { type: String, required: true },
//     dependencies: { type: [DependentNodeSchema], default: [] }
// });

const DependencyGraphSchema = new mongoose.Schema({
    files: String
});


const FileNodeSchema = new mongoose.Schema({
    name: { type: String, required: true },
    dependencies: { type: [DependentNodeSchema], default: [] } // Array instead of Set
});

const graphSchema = new mongoose.Schema({
    graphData: { type: Object, default: {} }, // Change to Object instead of Map
    projectId: { type: mongoose.Schema.Types.ObjectId, ref: 'Project', required: true }
}, { timestamps: true });


const FileNode = mongoose.model('FileNode', FileNodeSchema);
const UtilizedEntity = mongoose.model('UtilizedEntity', UtilizedEntitySchema);
const DependentNode = mongoose.model('DependentNode', DependentNodeSchema);
const DependencyGraph = mongoose.model('DependencyGraph', DependencyGraphSchema);
const Graph = mongoose.model('Graph', graphSchema);
module.exports =  { FileNode, UtilizedEntity, DependentNode, DependencyGraph, Graph };

File: api-gateway/app/mongo_models/DependencyGraph.py
from pydantic import BaseModel, Field
from typing import Optional, List, Dict
from datetime import datetime
from enum import Enum

class SourceEnum(str, Enum):
    Exporting = "Exporting"
    Importing = "Importing"

class UtilizedEntity(BaseModel):
    name: str
    type: str
    alias: Optional[str] = None
    source: SourceEnum

class DependentNode(BaseModel):
    name: str
    alias: Optional[str] = None
    valid: bool
    weight: List[UtilizedEntity] = Field(default_factory=list)

class FileNode(BaseModel):
    name: str
    dependencies: List[DependentNode] = Field(default_factory=list)

class DependencyGraph(BaseModel):
    files: Dict[str, FileNode] = Field(default_factory=dict)

# Input model for API calls, using projectTitle to find or create the project.
class GraphIn(BaseModel):
    projectTitle: str
    graphData: DependencyGraph